src/main/java/duke/shared/DateRange.java

```diff
@@ -0,0 +1,148 @@
+package duke.shared;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+
+import duke.constants.Constants;
+
+/**
+ * Encapsulates a pair of start datetime and end datetime.
+ */
+public class DateRange {
+    private LocalDateTime start;
+    private LocalDateTime end;
+
+    /**
+     * Creates a DateRange object with the given start and end datetime.
+     *
+     * @param start Start datetime.
+     * @param end   End datetime.
+     * @throws DukeException When start datetime occurs after end datetime.
+     */
+    public DateRange(LocalDateTime start, LocalDateTime end) throws DukeException {
+        this.start = start;
+        this.end = end;
+        if (isInvalid()) {
+            throw new InvalidDateException("Start date cannot be later than end date.");
+        }
+    }
+
+    /**
+     * Creates a DateRange object with the given start and end datetime strings.
+     *
+     * @param start String representing start datetime.
+     * @param end   String representing end datetime.
+     * @return DateRange object.
+     * @throws DukeException When start datetime occurs after end datetime.
+     */
+    public static DateRange createFromString(String start, String end) throws DukeException {
+        return new DateRange(LocalDateTime.parse(start, Constants.Input.DATETIME_FORMATTER),
+                LocalDateTime.parse(end, Constants.Input.DATETIME_FORMATTER));
+    }
+
+    /**
+     * Returns a DateRange object created from a single string representing the
+     * datetime range. The string can of the following formats: dd/MM/yyyy
+     * HHmm-dd/MM/yyyy HHmm or dd/MM/yyyy HHmm-HHmm The latter occurs on a single
+     * day.
+     *
+     * @param range String representing a range of dates.
+     */
+    public static DateRange createFromRange(String range) throws DukeException {
+        String[] rangeArr = range.split(Constants.Input.DATE_RANGE_SEPARATOR);
+        String start = rangeArr[0];
+        String datePart = rangeArr[0].split(" ")[0];
+        String end = rangeArr[1];
+        if (!end.contains(" ")) {
+            // Here end only has the time part, so we append the date part
+            end = String.format("%s %s", datePart, end);
+        }
+
+        return createFromString(start, end);
+    }
+
+    /**
+     * Returns whether the date range is valid.
+     *
+     * @return Whether the date range is valid.
+     */
+    public boolean isValid() {
```

Comment:
I wonder if having an isValid and isUnvalid function more confusing than just one of either? Just a personal thought haha

Reply:
Yep I agree it is quite unnecessary. Just my ruby-ist mindset kicking in :)